### ISSUE TITLE
Support External Consul Servers

### ISSUE THREAD

**Issue by lkysow** [hashicorp/consul-helm#253]
Support exposing Consul clients running in Kube to Consul servers running outside of Kube

**Comment by stvdilln** [comment]
I'm working on the PR doing some final testing.  I would like to fix as many scenarios as possible.

Do you see any flaws to this:
If we are only installing client (and not server) then client 8301,8302 are going to listen on NodePorts and can communicate with external server.

If we are only installing the server then 8300,8310,8302 are going to listen on NodePorts and this exposes the server cluster to outside world.

If we are installing both client and server:
The server listens on NodePorts, The client listens on Pod Ports.   When the server does a connect to the client it should open a socket on the pod port and the connection is pod port<->pod port.  When the client wants to connect it would be PodPort (client)<->NodePort (server).

If you want I can just fix the client connecting to an external server in this PR, but I think I can fix both the client to an external server and allowing external access to the server nodes running in K8s in one pass.

**Comment by lkysow** [comment]
> If you want I can just fix the client connecting to an external server in this PR, but I think I can fix both the client to an external server and allowing external access to the server nodes running in K8s in one pass.

We're interested only in the client to external server right now. We need to think more about the other cases.

Notes:
1. We don't want the clients listening on host ports (other than 8500 and 8502) by default, this should be optional
1. It should be a separate PR for exposing the servers. There's already some work done in https://github.com/hashicorp/consul-helm/pull/27.
1. If installing both client and server... we already support this so I wouldn't want to make any changes that aren't optional. We're not exposing the server on hostPorts right now because the default configuration has Server pods run on the same nodes as the client daemonset so there would be port conflicts.

**Comment by stvdilln** [comment]
I had started the work on doing an external server and it got messy quick so I abandoned that.  The possibilty of Client, or Server or Both running on same node quickly exploded the complexity.

I've submitted a PR that if you are not installing the server also, ports 8301 and 8302 are exposed via hostPorts so that the external consul cluster can connect to the k8s based clients.  Before this there was a one way communcation where the clients could connect to the external servers, but not the other way around.
The opening on 8301 and 8302 is contingent on server.enabled=false.
We can create a new explicit flag  client.useExternalServer or key the exposing of the ports to client.
We could also key off of client.join which indirectly implies an external server.  

Look though the PR 256 and let me know if you would like any changes.

**Comment by mekza** [comment]
@stvdilln Could you add a reference to your PR in this issue or a reference to this issue in your PR. It's going to be easier to keep track. Many thanks :)

**Comment by TylerReid** [comment]
@stvdilln I have been using this change in a test environment and it is working for me. Thanks for the PR! I was going to need to fork and do this myself. My setup is EKS nodes connecting to servers managed by https://github.com/hashicorp/terraform-aws-consul using cloud auto join.

**Comment by ilhaan** [comment]
I have tested this on a bare-metal cluster and it works very well. Thanks for the PR @stvdilln!

**Comment by adilyse** [comment]
Fixed by #289 which is a slight modification of #256.
